### ISSUE TITLE
Is it possible to run it in realtime?

### ISSUE THREAD

**Issue by Hasankanso** [nixingyang/AdaptiveReID#5]
if we suppose gallery_features do not change in runtime and we only get query images through the camera, then we extract features of the query, then we compare them using the provided metric? I believe it's similar to on_epoch_end method on solution.py

back to the original question, is it possible to run it in realtime? with webcam or ip camera stream

**Comment by nixingyang** [comment]
Yes. It is definitely feasible to develop a real-time demo for your use case.
In our experiments, the speed of the inference procedure using the ResNet50 backbone on single V100 32GB GPU is approximately 230 images per second. Naturally, the speed would be doubled if you set the `use_horizontal_flipping_in_evaluation` flag to `False`.
All in all, computing features for re-identification is fast on a decent GPU. No need to worry about that.

**Comment by Hasankanso** [comment]
Hello again,
I have a question regarding running solution.py in realtime.
extract_features method takes data_generator as argument
in real time I would be streaming images let's say through opencv.VideoCapture, which means I want to give single image for extract_features method.
another case is the gallery images, which is an array of images, probably around 20 images.

I'm struggling to adapt or even replace data_generator, it seems a bit more complicated than I thought.
Providing live demo would be great for this algorithm, instant access in order to benchmark different trained model.
Any hints/help are highly appreciated,
Thank you

**Comment by nixingyang** [comment]
Hi,
The challenge here is not running `solution.py` in realtime but implementing a separate script for the demo. Evidently, the `solution.py` script contains snippets for the training and evaluation procedures. However, many of those snippets are not needed in the demo, e.g., all the callbacks. Please find the pseudocode for your demo below:

- A generator which yields frames captured from the camera using OpenCV.
- A function which initializes the person detector and loads corresponding weights.
- A function which initializes the feature extractor (for re-identification) and loads corresponding weights. Use the command in the [Evaluation](https://github.com/nixingyang/AdaptiveReID#evaluation) section, and stop the execution at [Line 1111](https://github.com/nixingyang/AdaptiveReID/blob/274fd5fc4f8fc6ca90e2d18258b1d0845792fa41/solution.py#L1111). The `inference_model` variable at this point is the feature extractor.
- Once the program starts, compute the feature embeddings for the gallery set using the `predict_on_batch` method and save those to a file on disk. Make sure to preprocess the detected persons in a similar way to the [getitem](https://github.com/nixingyang/AdaptiveReID/blob/274fd5fc4f8fc6ca90e2d18258b1d0845792fa41/solution.py#L562-L584) method.
- Inside a for loop of the data generator, detect persons in the current frame and compute the feature embeddings for the query set. Find the closest match from the gallery set by comparing the cosine distances.

Please feel free to contact me if you need any further assistance.
All the best.
Xingyang

**Comment by Hasankanso** [comment]
I didn't have time to continue work on live scenario, now I'm fully dedicated for it. Thanks you for the pseudo code it helps a lot. I jumped to the third point since I can use images from dataset and test them. what I did:
1. load 3 images, 2 from same person and 1 different
2. calculated their features array using inference_model.predict
3. pass the results to compute_distance_matrix
here is what I tried:
```
 if evaluation_only:
        print("Freezing the whole model in the evaluation_only mode ...")
        training_model.trainable = False
        training_model.compile(**training_model.compile_kwargs)

        image1 = image.load_img('./Market-1501-v15.09.15/gt_bbox/0001_c2s1_001976_00.jpg', target_size= (384, 128, 3))
        image2 = image.load_img('./Market-1501-v15.09.15/bounding_box_test/0001_c5s3_032065_04.jpg', target_size= (384, 128, 3))
        red_guy = image.load_img('./Market-1501-v15.09.15/query/0004_c3s3_065619_00.jpg', target_size=(384, 128,3))

       #person1
        image1 = image.img_to_array(image1)
        image1 = np.expand_dims(image1, axis=0)
        result1 = inference_model.predict(image1)
        
       #person1
        image2 = image.img_to_array(image2)
        image2 = np.expand_dims(image2, axis=0)
        result2 = inference_model.predict(image2)
        
        #person2
        red_guy = image.img_to_array(red_guy)
        red_guy = np.expand_dims(red_guy, axis=0)
        red_guy = inference_model.predict(red_guy)

        result2 = np.array([result2, red_guy])
        result2 = result2.reshape(2, 4096)


        distance_matrix = compute_distance_matrix(query_image_features=result1, gallery_image_features=result2, metric="cosine", use_re_ranking= use_re_ranking)
        print(distance_matrix)
```
output:
`[[0.05850542 0.16167432]]`

the first value is lesser than the second one, it means image1 and image2 are the same person(which is correct) am I right? I have another question,  is there a threshold to know whether a person exists in the gallery or not? not just ranking them.
Thank you for your priceless assistance.

**Comment by nixingyang** [comment]
Hi,

There are some minor issues in your implementation:
- You are using functions from `tf.keras.preprocessing.image`. Although it should not differ significantly from the results using `OpenCV`, it is highly suggested to use the same code which is applied during training.
- It is necessary to apply `preprocess_input` to inputs.
- The first image is from the `gt_bbox` folder, and it looks suspicious. Please check the `image_folder_name` variable in [here](https://github.com/nixingyang/AdaptiveL2Regularization/blob/6279958976e25a128af8b40d7e8e77cb7484f828/datasets/market1501.py#L63-L70).

Hence, I am using the following snippets:
```python
if evaluation_only:
    print("Freezing the whole model in the evaluation_only mode ...")
    training_model.trainable = False
    training_model.compile(**training_model.compile_kwargs)

    root_folder_path = os.path.expanduser(
        "~/Documents/Local Storage/Dataset")
    dataset_folder_name = "Market-1501-v15.09.15"
    dataset_folder_path = os.path.join(root_folder_path,
                                        dataset_folder_name)

    image_file_path_list = [
        os.path.join(dataset_folder_path, item) for item in [
            "gt_bbox/0001_c2s1_001976_00.jpg",
            "bounding_box_test/0001_c5s3_032065_04.jpg",
            "query/0004_c3s3_065619_00.jpg"
        ]
    ]
    image_content_list = [
        read_image_file(image_file_path, input_shape)
        for image_file_path in image_file_path_list
    ]
    image_content_array = preprocess_input(np.array(image_content_list))

    first, second, third = inference_model.predict(image_content_array,
                                                    batch_size=batch_size)

    distance_matrix = test_evaluator_callback.compute_distance_matrix(
        query_image_features=[first],
        gallery_image_features=[second, third],
        metric="cosine",
        use_re_ranking=use_re_ranking)

    assert False, distance_matrix
```

With the `Market1501_ResNet50_9502037.h5` file from [Model Zoo](https://github.com/nixingyang/AdaptiveL2Regularization#model-zoo), the `distance_matrix` would be `[[0.06351125 0.35342598]]`. It means that the second image is more similar to the first image.

Modern person re-identification algorithms are designed to address a ranking problem. In order to make a binary prediction, you have to define a threshold somehow. One naive solution could be appending a dense layer, freezing the feature extractor, and optimizing the aforementioned dense layer by feeding positive/negative image pairs.

All the best.
Xingyang
